Add terraform and SSH  to pypeline

vittoriasalim opened the pull request:
- Add terraform 
- Add SSH

wonderyl commented on pypeline/benchmark.py:
```diff
@@ -19,40 +19,49 @@ def validate(self) -> list[Step]:
+    @abstractmethod
+    def get_credential_type(self) -> str:
+        pass
 @dataclass
```
Is this file formatted by a tool? It looks to me the format is not right.

wonderyl commented on pypeline/benchmark.py:
```diff
@@ -19,40 +19,49 @@ def validate(self) -> list[Step]:
     def login(self) -> list[Step]:
         pass
 
+    @abstractmethod
```
I wonder if you can move these fields as data fields of the Cloud class, then you don't need expose these getters anymore. I'm trying to avoid these boiler plates by using data class directly.

wonderyl commented on pypeline/resource/terraform/terraform.py:
```diff
@@ -0,0 +1,184 @@
+)
+
+
+class Terraform(Resource):
```
Could you make this a dataclass to avoid the boilerplate of init?

wonderyl replied:
Technically Terraform is not a resource, AKS is a resource. We use Terraform to set up an AKS cluster. I think Terraform could be a good base class that other resources inherit and reuse its implementations.

wonderyl commented on pypeline/resource/terraform/run_command.py:
```diff
@@ -0,0 +1,89 @@
+                "ARM_TENANT_ID": "$(AZURE_MI_TENANT_ID)",
+                "ARM_CLIENT_ID": "$(AZURE_MI_CLIENT_ID)",
+            }
+            if credential_type == "managed_identity"
```
Could you make credential type a Enum instead to avoid typo?

wonderyl commented on pypeline/cloud/azure.py:
```diff
@@ -86,3 +86,14 @@ def login(self) -> list[Step]:
+    def get_region(self) -> str:
+        return self.region
+
+    def get_credential_type(self) -> str:
```
Is this a property of the Cloud or a choice of the pipeline?

vittoriasalim replied:
property of cloud, will move it to cloud class to remove getters

wonderyl replied:
What I meant was, how you sign in is not a property of a Cloud class, it's a user's choice. You can pass it in as an option of login or other functions that use it.

wonderyl commented on pypeline/pipeline.py:
```diff
@@ -16,7 +16,11 @@ class Step:
     display_name: str = field(metadata={"yaml": "displayName"})
     script: str
-    env: Optional[dict[str, str]] = None
+    condition: Optional[str] = field(metadata={"yaml": "condition"}, default=None)
```
You only need to set metadata yaml when the output name is different than the variable name. E.g. the variable is display_name while the output yaml name is displayName.

wonderyl commented on pypeline/resource/terraform/input_variables.py:
```diff
@@ -0,0 +1,185 @@
+        "INPUT_VARIABLES": input_variables,
+        "DEBUG": "$(System.Debug)",
+    },
+    condition="ne(variables['SKIP_RESOURCE_MANAGEMENT'], 'true')",
```
Could we remove the env variable and make it an input to a resource? We should eliminate all env variables. It's really hard to track env var usage, and it cut though call stack without any trace.

vittoriasalim replied:
Right, Noted

wonderyl commented on pypeline/resource/terraform/input_variables.py:
```diff
@@ -0,0 +1,185 @@
+from textwrap import dedent
+from pipeline import Script
+
+set_input_variables_aws = lambda cloud, regions, input_variables={}: Script(
```
Could you make it a regular function? lambdas are usually used for short functions.

wonderyl replied:
Could you use typing please?

wonderyl commented on pypeline/resource/terraform/input_variables.py:
```diff
@@ -0,0 +1,185 @@
+from textwrap import dedent
+from pipeline import Script
+
+set_input_variables_aws = lambda cloud, regions, input_variables={}: Script(
```
Does it make sense to write aws, azure and gcp terraforms in 3 different classes?

vittoriasalim replied:
When I checked the Terraform Resource, the only place where we currently need separate commands for AWS, Azure, and GCP is in the set_input_variables function. 

I suppose for now, it’s best to keep them within Terraform for maintainability and to avoid high coupling—until it’s extended further. What do you think?

![image](https://github.com/user-attachments/assets/a822b633-da44-4ad9-a6aa-0ae51f6391f4)

wonderyl replied:
SGTM, let's how they are used in an actual resource to decide.

wonderyl commented on pypeline/resource/terraform/input_variables.py:
```diff
@@ -0,0 +1,185 @@
+        regional_config_str=$(echo $regional_config | jq -c .)
+        echo "Final regional config: $regional_config_str"
+        echo "##vso[task.setvariable variable=TERRAFORM_REGIONAL_CONFIG]$regional_config_str"
+    """
```
Should you align the """?

wonderyl replied:
Should you align the """?

wonderyl commented on pypeline/resource/terraform/input_variables.py:
```diff
@@ -0,0 +1,185 @@
+        "RUN_ID": "$(RUN_ID)",
+        "INPUT_VARIABLES": input_variables,
+        "DEBUG": "$(System.Debug)",
+        "SKU_TIER": "$(SKU_TIER)",
```
Ditto for these env vars. These should be input variables to a Resource.

wonderyl commented on pypeline/pipelines/controller/job_scheduling.py:
```diff
@@ -5,22 +5,27 @@
+            Python3(),
+            Terraform(
+                cloud_obj=cloud_az_eastus2
+            ),  # Dynamically retrieve cloud from layout
```
What does this comment mean?, are you talking about line 20?

wonderyl replied:
What does this comment mean?, are you talking about line 20?

wonderyl commented on pypeline/resource/terraform/terraform.py:
```diff
@@ -0,0 +1,184 @@
+
+
+class Terraform(Resource):
+    cloud_obj : Cloud = None
```
Should you just name it cloud? obj serves no meaning.

vittoriasalim replied:
Right!

vittoriasalim commented on pypeline/resource/terraform/run_command.py:
```diff
@@ -0,0 +1,89 @@
+        cd $TERRAFORM_WORKING_DIRECTORY
+
+        # Handle apply or destroy commands
+        if [[ "{command}" == "apply" || "{command}" == "destroy" ]]; then
```
Right. Better make it out of the script.

vittoriasalim commented:
@wonderyl 

Addition :
- Added SSH resource
- Set up a resource factory to avoid reinitializing shared properties for every resource.
- Moved all components to components.py to avoid circular imports and to keep all components in one place.
Corrections
- Removed environment variables and made them inputs to resources.
- Refactored Terraform into its own class, refactor run_command and input_variables
- Ensured proper typing and converted lambda expressions into functions.
- Used an Enum for credential types.
- Moved logic outside the script for better readability.
- Moved properties into data fields within the cloud class to avoid exposing getters and setters.

wonderyl commented on .gitignore:
```diff
@@ -48,6 +48,8 @@ terraform.*
 darwin-amd64/
 linux-amd64/
+# Exception: Do not ignore terraform.py
+!terraform.py
```
It's better to write line 48 more specific and avoid this exception.
E.g.
```
.terraform/
```
This will not ignore terraform.py

wonderyl commented on pypeline/components.py:
```diff
@@ -0,0 +1,58 @@
+
+@dataclass
+class Resource(ABC):
+    cloud: str
```
A resource can be a python binary, that's not related to Cloud at all.

wonderyl commented on pypeline/components.py:
```diff
@@ -0,0 +1,58 @@
+@dataclass
+class Resource(ABC):
+    cloud: str
+    regions: list[str]
```
Ditto

wonderyl commented on pypeline/components.py:
```diff
@@ -0,0 +1,58 @@
+
+from pipeline import Step
+
+# Keep components here to avoid circular imports
```
Could you elaborate more about this?

wonderyl commented on pypeline/components.py:
```diff
@@ -0,0 +1,58 @@
+# Factory class to create resources:
+# To avoid repeatedly define shared properties for every resource. usage: @ job_scheduling.py
+@dataclass
+class ResourceFactory:
```
This is not a factory pattern. https://en.wikipedia.org/wiki/Factory_method_pattern
As mentioned before, not all resource is associate with a cloud or a region.

wonderyl commented on pypeline/resource/setup.py:
```diff
@@ -1,96 +1,115 @@
+
+def set_run_id(run_id: str) -> Script:
+    if not run_id:
+        run_id = "$(Build.BuildId)-$(System.JobId)"
```
$(Build.BuildId) and $(System.JobId) azure pipeline's system env vars. We should not extract them out. In addition, it doesn't work, we don't define $(Build.BuildId) in Python.

wonderyl commented on pypeline/resource/setup.py:
```diff
@@ -1,96 +1,115 @@
 @dataclass
 class Setup(Resource):
     run_id: str
+    engine: str
```
Where is engine used?

wonderyl commented on pypeline/resource/setup.py:
```diff
@@ -1,96 +1,115 @@
     run_id: str
+    engine: str
     test_module_dir: str = ""
+    engine_input: dict = field(default_factory=dict)
```
Where is engine_input used?

wonderyl commented on pypeline/resource/setup.py:
```diff
@@ -1,96 +1,115 @@
+    script_modules_directory = f"$(Pipeline.Workspace)/s/{test_modules_dir}"
+
+    return Script(
+        display_name="Set Script Module Directory",
```
We still need to get `$(Pipeline.Workspace)` from the Azure pipeline env to get the full path of test_modules_directory. Then it sets a variable TEST_MODULES_DIR to record this value, which will be used in later steps. I don't think we can remove it.

wonderyl replied:
You need to distinguish 2 types of variables.
1 is used set in the env variable. e.g. `env={"TEST_MODULES_DIR": test_modules_dir},` we set this, so we can eliminate them. 
2 is `##vso[task.setvariable variable=TEST_MODULES_DIR]` these are azure pipeline variables, we can't remove them.

vittoriasalim replied:
> to record this value, which will be used in later steps. I don't think we can remove it.

Yes you are right!

wonderyl commented on pypeline/resource/setup.py:
```diff
@@ -1,96 +1,115 @@
     def validate(self) -> list[Step]:
+        return [validate_owner_info()]
+
+    def execute_tests(self) -> list[Step]:
```
Where is this method used?

vittoriasalim replied:
For validate_owner_info() function is called in the Layout during the resource validation step.

wonderyl replied:
The commend is on line 111. I was asking for `execute_tests`

vittoriasalim replied:
Right, it should have been removed. I was testing smthg and did remove it.

wonderyl commented on pypeline/cloud/azure.py:
```diff
@@ -1,99 +1,80 @@
-    region: str = "eastus"
+    cloud: str = "azure"
+    regions: list[str] = field(default_factory=lambda: ["eastus2"])
     subscription: str = os.getenv("AZURE_SUBSCRIPTION_ID")
```
I wonder why we need to get these vars from env vars, rather than pass them in when creating an Azure object?

vittoriasalim replied:
I checked some variables like AZURE_SUBSCRIPTION_ID, SERVICE_CONNECTION, etc.,  they appear to be set during the pipeline run—possibly for security reasons.
As for cloud, regions, and variables provided by user, we should be able to remove them in almost all scenarios.

wonderyl commented on pypeline/cloud/azure.py:
```diff
@@ -1,99 +1,80 @@
     subscription: str = os.getenv("AZURE_SUBSCRIPTION_ID")
     credential_type: CredentialType = CredentialType.SERVICE_CONNECTION
     azure_service_connection: str = os.getenv("AZURE_SERVICE_CONNECTION")
     azure_mi_client_id: str = os.getenv("AZURE_MI_CLIENT_ID")
```
Now we get the AZURE_MI_CLIENT_ID from the env we run Pypeline script. Obviously, it doesn't have it. I'm not 100% sure where is AZURE_MI_CLIENT_ID set, but I suspect it's from Azure pipeline, so we need to get the env var in the yaml file

wonderyl commented on pypeline/terraform/terraform.py:
```diff
@@ -0,0 +1,187 @@
+
+
+@dataclass
+class Terraform:
```
Should it inherit from the Resource class?

wonderyl replied:
Actually Terraform is not a resource, it's a tool. a AKS cluster is a resource, a resource group is a resource.

wonderyl commented on pypeline/benchmark.py:
```diff
@@ -1,67 +1,61 @@
             steps=self.setup.setup()
             + self.cloud.login()
             + [step for r in self.resources for step in r.setup()]
+            + self.terraform.setup()
```
You should model Terraform as a resource and add it to self.resources.

wonderyl commented on pypeline/benchmark.py:
```diff
@@ -1,67 +1,61 @@
+            + self.terraform.setup()
+            + [
+                self.terraform.create_resource_group(),
+                self.terraform.run_command(command="version"),
```
It's better to have version, init, apply as separate functions or at least make command an enum, so that it doesn't take arbitrary string.
